File: app/jobs.py
"""Single-worker background queue driving the transcription pipeline.

The SessionStore is the source of truth for status/results; this queue only
runs work on a serialized executor (max_workers=1, so CPU isn't oversubscribed
and uploads queue) and records state transitions on the store.
"""

from __future__ import annotations

import logging
import threading
from concurrent.futures import ThreadPoolExecutor
from typing import Callable

logger = logging.getLogger(__name__)


class Cancelled(Exception):
    """Raised inside the pipeline thread when a cancel() call is received."""


class JobQueue:
    def __init__(self, store, run_session: Callable[[str, threading.Event], None], broker=None):
        """run_session(session_id, cancel_event): execute the pipeline and persist artifacts.

        It must call store.mark_done(...) itself (it has the result metadata);
        the queue handles mark_running and error capture. When a ``broker`` is
        given, terminal status transitions are published to SSE subscribers
        *after* the store is updated (so a client reacting to the event reads a
        consistent row).
        """
        self._store = store
        self._run_session = run_session
        self._broker = broker
        self._executor = ThreadPoolExecutor(max_workers=1)
        self._lock = threading.Lock()
        self._cancel_events: dict[str, threading.Event] = {}

    def submit(self, session_id: str) -> None:
        event = threading.Event()
        with self._lock:
            self._cancel_events[session_id] = event
        self._executor.submit(self._run, session_id, event)

    def cancel(self, session_id: str) -> None:
        """Signal the running job to stop at the next stage boundary."""
        with self._lock:
            event = self._cancel_events.get(session_id)
        if event is not None:
            event.set()

    def shutdown(self) -> None:
        """Stop accepting work on app quit. ``wait=False`` so a long-running job
        doesn't block teardown; an interrupted job is reconciled to ``error`` by
        ``SessionStore.reconcile_startup`` on the next boot."""
        self._executor.shutdown(wait=False)

    def _publish(self, session_id: str, event: dict) -> None:
        if self._broker is not None:
            self._broker.publish(session_id, event)

    def _run(self, session_id: str, cancel_event: threading.Event) -> None:
        self._store.mark_running(session_id)
        try:
            self._run_session(session_id, cancel_event)
        except Cancelled:
            logger.info("Session %s cancelled", session_id)
            # Don't mark_error — the delete endpoint already removed the row.
        except Exception as exc:  # noqa: BLE001 - surface any failure to the UI
            logger.exception("Session %s failed", session_id)
            self._store.mark_error(session_id, str(exc))
            self._publish(session_id, {"status": "error"})
        else:
            self._publish(session_id, {"status": "done"})
        finally:
            with self._lock:
                self._cancel_events.pop(session_id, None)

File: adapters/server/jobs/jobs.cpp
#include "jobs/jobs.hpp"

#include <nlohmann/json.hpp>

#include "db/session_store.hpp"
#include "log/log.hpp"
#include "sse/broker.hpp"

using nlohmann::json;

namespace whisperx::server::jobs {

JobQueue::JobQueue(whisperx::db::SessionStore& store, RunSession run_session,
                   whisperx::server::sse::Broker* broker)
    : store_(store),
      run_session_(std::move(run_session)),
      broker_(broker),
      worker_([this] { worker_loop(); }) {}

JobQueue::~JobQueue() { shutdown(); }

void JobQueue::submit(const std::string& session_id) {
    {
        std::lock_guard<std::mutex> lk(mu_);
        if (stop_) return;
        cancel_flags_[session_id] = std::make_shared<std::atomic<bool>>(false);
        queue_.push_back(session_id);
    }
    cv_.notify_one();
}

void JobQueue::cancel(const std::string& session_id) {
    std::lock_guard<std::mutex> lk(mu_);
    auto it = cancel_flags_.find(session_id);
    if (it != cancel_flags_.end()) it->second->store(true);
}

void JobQueue::shutdown() {
    {
        std::lock_guard<std::mutex> lk(mu_);
        if (stop_) return;
        stop_ = true;
    }
    cv_.notify_all();
    if (worker_.joinable()) worker_.join();
}

void JobQueue::worker_loop() {
    for (;;) {
        std::string id;
        {
            std::unique_lock<std::mutex> lk(mu_);
            cv_.wait(lk, [this] { return stop_ || !queue_.empty(); });
            if (stop_ && queue_.empty()) return;
            id = std::move(queue_.front());
            queue_.pop_front();
        }
        run_one(id);
    }
}

void JobQueue::run_one(const std::string& session_id) {
    auto logger = whisperx::server::log::get("jobs");
    CancelFlag cancel;
    {
        std::lock_guard<std::mutex> lk(mu_);
        cancel = cancel_flags_[session_id];
    }

    store_.mark_running(session_id);
    bool ok = false;
    bool cancelled = false;
    try {
        run_session_(session_id, cancel);
        ok = true;
    } catch (const Cancelled&) {
        // The delete endpoint already removed the row — don't mark_error.
        cancelled = true;
        logger->info("Session {} cancelled", session_id);
    } catch (const std::exception& exc) {
        logger->error("Session {} failed: {}", session_id, exc.what());
        store_.mark_error(session_id, exc.what());
    } catch (...) {
        logger->error("Session {} failed: unknown error", session_id);
        store_.mark_error(session_id, "unknown error");
    }

    // Terminal SSE *after* the store row is updated (consistent read for a
    // client reacting to the event). Cancelled jobs publish nothing.
    if (broker_ != nullptr && !cancelled)
        broker_->publish(session_id,
                         json{{"status", ok ? "done" : "error"}});

    std::lock_guard<std::mutex> lk(mu_);
    cancel_flags_.erase(session_id);
}

}  // namespace whisperx::server::jobs
